### PR TITLE
feat: return reasons and remove from args. 

### PR DESCRIPTION
**Pull request by msohailhussain**
## Summary
- returning reasons and removed from args
- added overloaded operator in DecisionReasons
- Now only one class is using DecisionReasons, toReport filters based on the DecisionOptions.

## Test plan
- All unit tests and FSC must pass.

### CODE REVIEW

File: OptimizelySDK/Bucketing/DecisionService.cs [hunk 1]
```diff
@@ -485,13 +481,16 @@ public void SaveVariation(Experiment experiment, Variation variation, UserProfil
...
+            var userMeetConditionsResultEveryoneElse = ExperimentUtils.DoesUserMeetAudienceConditions(config, everyoneElseRolloutRule, filteredAttributes, LOGGING_KEY_TYPE_RULE, "Everyone Else", Logger);
+            reasons += userMeetConditionsResultEveryoneElse.DecisionReasons;
+            if (userMeetConditionsResultEveryoneElse.ResultObject)
             {
-                variation = Bucketer.Bucket(config, everyoneElseRolloutRule, bucketingId, userId, reasons);
-                if (variation != null && !string.IsNullOrEmpty(variation.Id))
+                variationResult = Bucketer.Bucket(config, everyoneElseRolloutRule, bucketingIdResult.ResultObject, userId);
+                if (!string.IsNullOrEmpty(variationResult?.ResultObject?.Id))
                 {
+                    reasons += variationResult.DecisionReasons;
```
Comment (jaeopt) [inline OptimizelySDK/Bucketing/DecisionService.cs:491]: Move this line to before line 489

File: OptimizelySDK/Optimizely.cs [hunk 2]
```diff
@@ -437,7 +437,7 @@ public Variation GetForcedVariation(string experimentKey, string userId)
             if (!ValidateStringInputs(inputValues))
                 return null;
 
-            return DecisionService.GetForcedVariation(experimentKey, userId, config);
+            return DecisionService.GetForcedVariation(experimentKey, userId, config)?.ResultObject;
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:440]: remove "?". Result never returns null.

File: OptimizelySDK/Optimizely.cs [hunk 3]
```diff
@@ -809,17 +808,18 @@ public OptimizelyJSON GetFeatureVariableJSON(string featureKey, string variableK
             
             var optimizelyJSON = new OptimizelyJSON(variableMap, ErrorHandler, Logger);
 
-            var decisionSource = flagDecision?.Source ?? FeatureDecision.DECISION_SOURCE_ROLLOUT;
+            var decisionSource = flagDecisionResult?.ResultObject?.Source ?? FeatureDecision.DECISION_SOURCE_ROLLOUT;
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:811]:
```suggestion
            var decisionSource = flagDecisionResult.ResultObject?.Source ?? FeatureDecision.DECISION_SOURCE_ROLLOUT;
```

File: OptimizelySDK/Optimizely.cs [hunk 4]
```diff
@@ -1072,18 +1072,18 @@ public List<string> GetEnabledFeatures(string userId, UserAttributes userAttribu
                 valuesMap.Add(featureVariable.Key, typeCastedValue);
             }
             var sourceInfo = new Dictionary<string, string>();
-            if (decision?.Source == FeatureDecision.DECISION_SOURCE_FEATURE_TEST)
+            if (decisionResult?.ResultObject?.Source == FeatureDecision.DECISION_SOURCE_FEATURE_TEST)
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:1075]:
```suggestion
            if (decisionResult.ResultObject?.Source == FeatureDecision.DECISION_SOURCE_FEATURE_TEST)
```

File: OptimizelySDK/Optimizely.cs [hunk 5]
```diff
@@ -1072,18 +1072,18 @@ public List<string> GetEnabledFeatures(string userId, UserAttributes userAttribu
...
+                sourceInfo["variationKey"] = decisionResult.ResultObject.Variation.Key;
             }
 
             var decisionInfo = new Dictionary<string, object>
             {
                 { "featureKey", featureKey },
                 { "featureEnabled", featureEnabled },
                 { "variableValues", valuesMap },
-                { "source", decision?.Source },
+                { "source", decisionResult?.ResultObject?.Source },
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:1086]:
```suggestion
                { "source", decisionResult.ResultObject?.Source },
```

**Review by jaeopt — CHANGES_REQUESTED** [review]
Overall looks good. 
I suggest to add tests validating core messages of reasons returned from decision to make sure reason collecting works as expected. See my comments.

File: OptimizelySDK/Bucketing/DecisionService.cs [hunk 6]
```diff
@@ -99,23 +99,32 @@ public DecisionService(Bucketer bucketer, IErrorHandler errorHandler, UserProfil
...
+            {
+                return forcedVariationResult.SetReasons(reasons);
+            }
 
-            var variation = GetWhitelistedVariation(experiment, userId, reasons);
+            var variationResult = GetWhitelistedVariation(experiment, userId);
+            reasons += variationResult.DecisionReasons;
 
-            if (variation != null) return variation;
+            if (variationResult?.ResultObject != null)
```
[the remark this answers is not in the document]
Comment (msohailhussain) [reply OptimizelySDK/Bucketing/DecisionService.cs:124]: yes , true.

File: OptimizelySDK/Bucketing/DecisionService.cs [hunk 7]
```diff
@@ -146,31 +156,34 @@ public DecisionService(Bucketer bucketer, IErrorHandler errorHandler, UserProfil
...
-                string bucketingId = GetBucketingId(userId, filteredAttributes, reasons);
+                var bucketingIdResult = GetBucketingId(userId, filteredAttributes);
+                reasons += bucketingIdResult.DecisionReasons;
 
-                variation = Bucketer.Bucket(config, experiment, bucketingId, userId, reasons);
+                variationResult = Bucketer.Bucket(config, experiment, bucketingIdResult.ResultObject, userId);
+                reasons += variationResult.DecisionReasons;
 
-                if (variation != null && variation.Key != null)
+                if (variationResult != null && variationResult.ResultObject?.Key != null)
```
[the remark this answers is not in the document]
Comment (msohailhussain) [reply OptimizelySDK/Bucketing/DecisionService.cs:170]: +1

File: OptimizelySDK/Bucketing/DecisionService.cs [hunk 8]
```diff
@@ -461,20 +461,15 @@ public void SaveVariation(Experiment experiment, Variation variation, UserProfil
                 if (userMeetConditionsResult.ResultObject)
                 {
                     variationResult = Bucketer.Bucket(config, rolloutRule, bucketingIdResult.ResultObject, userId);
-
+                    reasons += variationResult?.DecisionReasons;
```
Comment (jaeopt) [inline OptimizelySDK/Bucketing/DecisionService.cs:464]:
```suggestion
                    reasons += variationResult.DecisionReasons;
```

File: OptimizelySDK/Bucketing/DecisionService.cs [hunk 9]
```diff
@@ -486,9 +481,10 @@ public void SaveVariation(Experiment experiment, Variation variation, UserProfil
             if (userMeetConditionsResultEveryoneElse.ResultObject)
             {
                 variationResult = Bucketer.Bucket(config, everyoneElseRolloutRule, bucketingIdResult.ResultObject, userId);
+                reasons += variationResult?.DecisionReasons;
```
Comment (jaeopt) [inline OptimizelySDK/Bucketing/DecisionService.cs:484]:
```suggestion
                reasons += variationResult.DecisionReasons;
```

**Review by jaeopt — APPROVED** [review]
LGTM with nits

**Pull request merged**